[feat] tracker url plugin: use ClearURL tracking param list

Bnyro opened the pull request:
## What does this PR do?
- this PR implements a (for us) automated way of having up to date tracking rules using ClearURLs rules

## Why is this change important?
- see #4472

## How to test this PR locally?
- enable the tracker url remover plugin and do a search for something with tracker params
- `dan brown amazon`  has some for example
- compare result urls to the urls when the plugin is disabled

## Author's checklist
- we could also add other sources to `update_tracker_patterns.py` and merge the tracking rules, e.g. from AdGuard, with the existing ones
- The format used is quite simple and thus should be usable with other sources too (if we parse them to this format) 

## Related issues
closes #4472

Bnyro committed:
[feat] tracker url plugin: use ClearURL tracking param list

Bnyro committed:
[data] update searx.data - add tracker_patterns.json

return42 committed:
[mod] plugins: activate tracker url plugin by default

With the new implementation for removing tracker arguments from URLs, we use the
tracker definitions of the ClearURLs project [1].  These definitions are
regularly updated and significantly improve data protection, which is why we
should now also enable this plugin by default.

[1] https://github.com/ClearURLs

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 commented on searx/plugins/tracker_url_remover.py:
```diff
@@ -39,20 +36,53 @@ def __init__(self, plg_cfg: "PluginCfg") -> None:
+    def on_result(self, request: "SXNG_Request", search: "SearchWithPlugins", result: Result) -> bool:
+
+        result.filter_urls(self.filter_url_field)
+        return True
```
To remove tracker arguments from *all* URLs (including links in the infobox and other areas) we have to use the [`Result.filter_urls(..)`](https://docs.searxng.org/dev/result_types/base_result.html#searx.result_types._base.Result.filter_urls) method / its the same procedure we have in the hostname-replace plugin.

return42 commented on searxng_extra/update/update_tracker_patterns.py:
```diff
@@ -0,0 +1,36 @@
+                "exceptions": [exc.replace("\\\\", "\\") for exc in rule["exceptions"]],
+                "trackerParams": rule["rules"],
+            }
+        )
```
When PR https://github.com/searxng/searxng/pull/4836 has been merged we have a boilerplate to manage `data` in a *ExpireCache*.

.. and instead of the json file which needs to be created today, we can directly read the rules from `CLEAR_LIST_URL` in the *ExpireCahe* .. 

ping @dalf: Here's an example where SearXNG uses `data` from external projects. To ensure efficient access to the `data` objects, these objects are loaded into an (expiring) cache.  This cache is not yet implemented in this PR, but its related to what we discuss in #4836.

return42 committed:
[mod] data: implement a simple tracker URL (SQL) database

On demand, the tracker data is loaded directly into the cache, so that the
maintenance of this data via PRs is no longer necessary.

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

unixfox commented on searx/settings.yml:
```diff
@@ -245,7 +245,7 @@ plugins:
 
   searx.plugins.tracker_url_remover.SXNGPlugin:
-    active: false
+    active: true
```
I think this shouldn't be by default because this may have potential side effects that we do not know.

And we do not maintain the list in "CLEAR_LIST_URL".

It's like hosting a pihole or adguard home, you start to add multiple sources but then some websites starts to break or "half break".

The user should be aware of such feature by turning it on by themselves.

return42 replied:
> The user should be aware of such a feature by turning it on by themselves.

Under our promise "Users are neither tracked nor profiled," I, as a user, would assume that all links offered to me are also cleared of trackers.

> And we do not maintain the list in "CLEAR_LIST_URL."

If problems arise, the user can disable URL cleanup.

The lists come from this project: https://docs.clearurls.xyz/1.27.3/ . It can be assumed that they are designed to be conservative; otherwise, there would be more issue reports in the project.

unixfox replied:
Yes, this plugin is great for privacy. But I would be in more confidence to enable it by default if that was a list maintained by us or a fixed algorithm written by us.

But that's not the case. I would rather by default present to the user the original unmodified link than having some weird edge cases where the link is broken due to this plugin. Privacy is important, but what's more important is usability. The majority of people use a software because it works great for them and privacy is a nice perk for them. If the software partially works, they are not going to use it, even though it is great for privacy.

Since, this plugin can be easily activated in the settings page, I do not see the point of enabling it by default.

unixfox replied:
> It can be assumed that they are designed to be conservative; otherwise, there would be more issue reports in the project.

Look at the issues in https://github.com/ClearURLs/Rules/issues. I found one issue about those "edge cases" that could happen in SearXNG:
- https://github.com/ClearURLs/Rules/issues/129

Not saying that we are going to encounter a lot of issues like these. But on a large scale of users, the likeliness is going to be higher. That's why it's better to be opt-in, so that the user is aware of such possible breakage.

And this list was primarily used for the ClearURLs extension, not for SearXNG. I already had some bad experience in the past by trying to use an adblock list URLs in pihole, that lead to some website not functioning anymore because the regex was designed for adblock not pihole. Adblock acts on the browser level, whereas pihole acts on the DNS level.

return42 replied:
It is only logical that external data can also contain errors. There are errors here and in many other places where we get data from external sources (and as an aggregator, all we do is collect and process external data).

There's no right or wrong answer to the question of whether this plugin should be enabled by default or not. I'd just like to see in practice how it behaves overall if we enable it by default.

If we later realize that this was a bad decision because there are too many errors, then it's easy to set the default to "off"... at least then we know for sure why "off" is the better choice by default.

#### TLDR: *Let's just try it out and make fewer assumptions.*  ..

unixfox replied:
The right path is to let people try a new feature in a testing phase and then enable it after if the general public wants to have it enabled by default.

Not the reverse. Otherwise, you impact everyone, which is not a good choice when testing a new feature.

return42 replied:
I give up asking for a little more relaxation and hope that my conversation partner is willing to follow the recommendations or requests of others.

Bnyro replied:
> And this list was primarily used for the ClearURLs extension, not for SearXNG. I already had some bad experience in the past by trying to use an adblock list URLs in pihole, that lead to some website not functioning anymore because the regex was designed for adblock not pihole. Adblock acts on the browser level, whereas pihole acts on the DNS level.
>
ClearURL is web based, so we don't have such an issue.

> There's no right or wrong answer to the question of whether this plugin should be enabled by default or not. I'd just like to see in practice how it behaves overall if we enable it by default.
>
I think it's a good compromise to enable it by default and disable it if there are many complaints, we'll do it like that 👍

unixfox replied:
Marking my veto on this. I can provide some efforts in publicly sharing this improvement for asking people to test it. But I do not want this feature to be on by default, for all the reasons explained above.

Bnyro replied:
Well, we have a 2:1 vote here and you didn't seem to care about our vetos in other PRs either, so we'll do it as we said.
